File: core-python/src/rappel_core/backends/postgres.py
"""Postgres backend for persisting runner state and action results."""

from __future__ import annotations

import asyncio
import contextlib
import dataclasses
import pickle
import sys
from collections import Counter
from threading import Lock
from typing import Any, Iterator, Sequence
from uuid import uuid4

from psycopg_pool import ConnectionPool

from .base import ActionDone, BaseBackend, GraphUpdate, InstanceDone, QueuedInstance

DEFAULT_DSN = "postgresql://rappel:rappel@localhost:5432/rappel_core"


def _ensure_proto_aliases() -> None:
    """Ensure protobuf modules are importable by their generated names."""
    if "ast_pb2" not in sys.modules:
        try:
            from proto import ast_pb2 as proto_ast
        except ImportError:
            proto_ast = None
        if proto_ast is not None:
            sys.modules["ast_pb2"] = proto_ast
    if "messages_pb2" not in sys.modules:
        try:
            from proto import messages_pb2 as proto_messages
        except ImportError:
            proto_messages = None
        if proto_messages is not None:
            sys.modules["messages_pb2"] = proto_messages


class PostgresBackend(BaseBackend):
    """Persist runner state and action results in Postgres."""

    _pool: ConnectionPool | None = None
    _pool_dsn: str | None = None
    _pool_lock = Lock()
    _schema_ready = False
    _schema_lock = Lock()

    def __init__(self, dsn: str = DEFAULT_DSN) -> None:
        _ensure_proto_aliases()
        self._dsn = dsn
        self._pool = self._get_pool(dsn)
        self._query_counts: Counter[str] = Counter()
        self._query_count_lock = Lock()
        self._batch_size_counts: dict[str, Counter[int]] = {}
        self._batch_size_lock = Lock()
        self._ensure_schema()

    def batching(self) -> contextlib.AbstractContextManager[BaseBackend]:
        return _postgres_batch(self)

    def save_graphs(self, graphs: Sequence[GraphUpdate]) -> None:
        if not graphs:
            return
        with self._pool.connection() as conn:
            self._update_graphs_in(conn, graphs)

    def save_actions_done(self, actions: Sequence[ActionDone]) -> None:
        if not actions:
            return
        payloads = [
            (
                action.node_id,
                action.action_name,
                action.attempt,
                self._serialize(action.result),
            )
            for action in actions
        ]
        self._copy_rows(
            "COPY runner_actions_done (node_id, action_name, attempt, result) FROM STDIN",
            payloads,
        )

    def save_instances_done(self, instances: Sequence[InstanceDone]) -> None:
        if not instances:
            return
        with self._pool.connection() as conn:
            self._update_instances_done_in(conn, instances)

    async def get_queued_instances(self, size: int) -> list[QueuedInstance]:
        if size <= 0:
            return []
        return await asyncio.to_thread(self._fetch_queued_instances, size)

    def queue_instances(self, instances: Sequence[QueuedInstance]) -> None:
        """Insert queued instances for run-loop consumption."""
        if not instances:
            return
        payloads: list[tuple[Any, Any]] = []
        instance_payloads: list[tuple[Any, Any, Any]] = []
        for instance in instances:
            instance_id = instance.instance_id
            if instance_id is None:
                instance_id = uuid4()
                instance = dataclasses.replace(instance, instance_id=instance_id)
            payloads.append((instance_id, self._serialize(instance)))
            graph = self._build_graph_update(instance, instance_id)
            instance_payloads.append(
                (
                    instance_id,
                    instance.entry_node,
                    self._serialize(graph),
                )
            )
        with self._pool.connection() as conn:
            with conn.transaction():
                self._copy_rows_in(
                    conn,
                    "COPY queued_instances (instance_id, payload) FROM STDIN",
                    payloads,
                )
                self._copy_rows_in(
                    conn,
                    "COPY runner_instances (instance_id, entry_node, state) FROM STDIN",
                    instance_payloads,
                )

    def clear_queue(self) -> None:
        """Delete all queued instances from the backing table."""
        with self._pool.connection() as conn:
            with conn.cursor() as cur:
                self._count_query("delete:queued_instances_all")
                cur.execute("DELETE FROM queued_instances")

    def clear_all(self) -> None:
        """Delete all persisted runner data for a clean benchmark run."""
        with self._pool.connection() as conn:
            with conn.cursor() as cur:
                self._count_query("truncate:runner_tables")
                cur.execute(
                    """
                    TRUNCATE runner_graph_updates,
                             runner_actions_done,
                             runner_instances,
                             runner_instances_done,
                             queued_instances
                    RESTART IDENTITY
                    """
                )

    def _fetch_queued_instances(self, size: int) -> list[QueuedInstance]:
        if size <= 0:
            return []
        with self._pool.connection() as conn:
            with conn.transaction():
                with conn.cursor() as cur:
                    self._count_query("select:queued_instances")
                    cur.execute(
                        """
                        SELECT instance_id, payload
                        FROM queued_instances
                        ORDER BY created_at
                        LIMIT %s
                        FOR UPDATE SKIP LOCKED
                        """,
                        (size,),
                    )
                    rows = cur.fetchall()
                    if not rows:
                        return []
                    self._count_batch_size("select:queued_instances", len(rows))
                    instance_ids = [row[0] for row in rows]
                    self._count_query("delete:queued_instances_by_id")
                    cur.execute(
                        "DELETE FROM queued_instances WHERE instance_id = ANY(%s)",
                        (instance_ids,),
                    )
        instances: list[QueuedInstance] = []
        for instance_id, payload in rows:
            instance = self._deserialize(payload)
            if instance.instance_id != instance_id:
                instance = dataclasses.replace(instance, instance_id=instance_id)
            instances.append(instance)
        return instances

    def _ensure_schema(self) -> None:
        if self.__class__._schema_ready:
            return
        with self.__class__._schema_lock:
            if self.__class__._schema_ready:
                return
            with self._pool.connection() as conn:
                with conn.cursor() as cur:
                    self._count_query("schema:runner_graph_updates")
                    cur.execute(
                        """
                        CREATE TABLE IF NOT EXISTS runner_graph_updates (
                            id BIGSERIAL PRIMARY KEY,
                            created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                            state BYTEA NOT NULL
                        )
                        """
                    )
                    self._count_query("schema:runner_actions_done")
                    cur.execute(
                        """
                        CREATE TABLE IF NOT EXISTS runner_actions_done (
                            id BIGSERIAL PRIMARY KEY,
                            created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                            node_id UUID NOT NULL,
                            action_name TEXT NOT NULL,
                            attempt INTEGER NOT NULL,
                            result BYTEA
                        )
                        """
                    )
                    self._count_query("schema:runner_instances")
                    cur.execute(
                        """
                        CREATE TABLE IF NOT EXISTS runner_instances (
                            instance_id UUID PRIMARY KEY,
                            created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                            entry_node UUID NOT NULL,
                            state BYTEA,
                            result BYTEA,
                            error BYTEA
                        )
                        """
                    )
                    self._count_query("schema:runner_instances_done")
                    cur.execute(
                        """
                        CREATE TABLE IF NOT EXISTS runner_instances_done (
                            id BIGSERIAL PRIMARY KEY,
                            created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                            executor_id UUID NOT NULL,
                            entry_node UUID NOT NULL,
                            result BYTEA,
                            error BYTEA
                        )
                        """
                    )
                    self._count_query("schema:queued_instances")
                    cur.execute(
                        """
                        CREATE TABLE IF NOT EXISTS queued_instances (
                            instance_id UUID PRIMARY KEY,
                            created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
                            payload BYTEA NOT NULL
                        )
                        """
                    )
            self.__class__._schema_ready = True

    def _copy_rows(self, query: str, rows: Sequence[Sequence[Any]]) -> None:
        if not rows:
            return
        with self._pool.connection() as conn:
            self._copy_rows_in(conn, query, rows)

    def _copy_rows_in(self, conn: Any, query: str, rows: Sequence[Sequence[Any]]) -> None:
        if not rows:
            return
        self._count_query(self._copy_label(query))
        self._count_batch_size(self._copy_label(query), len(rows))
        with conn.cursor() as cur:
            with cur.copy(query) as copy:
                for row in rows:
                    copy.write_row(row)

    def _update_graphs_in(self, conn: Any, graphs: Sequence[GraphUpdate]) -> None:
        if not graphs:
            return
        self._count_query("update:runner_instances_state")
        self._count_batch_size("update:runner_instances_state", len(graphs))
        payloads = [(self._serialize(graph), graph.instance_id) for graph in graphs]
        with conn.cursor() as cur:
            cur.executemany(
                """
                UPDATE runner_instances
                SET state = %s
                WHERE instance_id = %s
                """,
                payloads,
            )

    def _update_instances_done_in(
        self, conn: Any, instances: Sequence[InstanceDone]
    ) -> None:
        if not instances:
            return
        self._count_query("update:runner_instances_result")
        self._count_batch_size("update:runner_instances_result", len(instances))
        payloads = [
            (
                self._serialize_optional(instance.result),
                self._serialize_optional(instance.error),
                instance.executor_id,
            )
            for instance in instances
        ]
        with conn.cursor() as cur:
            cur.executemany(
                """
                UPDATE runner_instances
                SET result = %s, error = %s
                WHERE instance_id = %s
                """,
                payloads,
            )

    def _build_graph_update(self, instance: QueuedInstance, instance_id: UUID) -> GraphUpdate:
        if instance.state is not None:
            nodes = instance.state.nodes
            edges = instance.state.edges
        else:
            nodes = instance.nodes
            edges = instance.edges
        if nodes is None or edges is None:
            raise ValueError("queued instance missing runner state for persistence")
        return GraphUpdate(instance_id=instance_id, nodes=nodes, edges=edges)

    @classmethod
    def _get_pool(cls, dsn: str) -> ConnectionPool:
        with cls._pool_lock:
            if cls._pool is None:
                cls._pool = ConnectionPool(conninfo=dsn, min_size=1, max_size=10)
                cls._pool_dsn = dsn
            elif cls._pool_dsn != dsn:
                raise ValueError("PostgresBackend DSN mismatch with global pool")
            return cls._pool

    @staticmethod
    def _serialize(value: Any) -> bytes:
        _ensure_proto_aliases()
        return pickle.dumps(value, protocol=pickle.HIGHEST_PROTOCOL)

    @classmethod
    def _serialize_optional(cls, value: Any | None) -> bytes | None:
        if value is None:
            return None
        return cls._serialize(value)

    @staticmethod
    def _deserialize(payload: bytes | memoryview) -> QueuedInstance:
        _ensure_proto_aliases()
        if isinstance(payload, memoryview):
            payload = payload.tobytes()
        value = pickle.loads(payload)
        if not isinstance(value, QueuedInstance):
            raise TypeError("queued instance payload did not decode to QueuedInstance")
        return value

    def _count_query(self, label: str) -> None:
        with self._query_count_lock:
            self._query_counts[label] += 1

    def _count_batch_size(self, label: str, size: int) -> None:
        if size <= 0:
            return
        with self._batch_size_lock:
            counter = self._batch_size_counts.get(label)
            if counter is None:
                counter = Counter()
                self._batch_size_counts[label] = counter
            counter[size] += 1

    def _copy_label(self, query: str) -> str:
        parts = query.strip().split()
        if len(parts) >= 2:
            return f"copy:{parts[1]}"
        return "copy:unknown"

    def query_counts(self) -> dict[str, int]:
        with self._query_count_lock:
            return dict(self._query_counts)

    def batch_size_counts(self) -> dict[str, dict[int, int]]:
        with self._batch_size_lock:
            return {label: dict(counter) for label, counter in self._batch_size_counts.items()}


class _PostgresBatch(BaseBackend):
    def __init__(self, backend: PostgresBackend, conn: Any) -> None:
        self._backend = backend
        self._conn = conn

    def save_graphs(self, graphs: Sequence[GraphUpdate]) -> None:
        if not graphs:
            return
        self._backend._update_graphs_in(self._conn, graphs)

    def save_actions_done(self, actions: Sequence[ActionDone]) -> None:
        if not actions:
            return
        payloads = [
            (
                action.node_id,
                action.action_name,
                action.attempt,
                self._backend._serialize(action.result),
            )
            for action in actions
        ]
        self._backend._copy_rows_in(
            self._conn,
            "COPY runner_actions_done (node_id, action_name, attempt, result) FROM STDIN",
            payloads,
        )

    def save_instances_done(self, instances: Sequence[InstanceDone]) -> None:
        if not instances:
            return
        self._backend._update_instances_done_in(self._conn, instances)

    async def get_queued_instances(self, size: int) -> list[QueuedInstance]:
        return await self._backend.get_queued_instances(size)


@contextlib.contextmanager
def _postgres_batch(backend: PostgresBackend) -> Iterator[_PostgresBatch]:
    pool = backend._pool
    if pool is None:
        raise RuntimeError("Postgres connection pool not initialized")
    with pool.connection() as conn:
        with conn.transaction():
            backend._count_query("batch:transaction")
            yield _PostgresBatch(backend, conn)

File: src/rappel_core/backends/postgres.rs
//! Postgres backend for persisting runner state and action results.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use futures::future::BoxFuture;
use sqlx::{PgPool, Postgres, QueryBuilder, Row};
use uuid::Uuid;

use crate::db;
use crate::observability::obs;

use super::base::{
    ActionDone, BackendError, BackendResult, BaseBackend, GraphUpdate, InstanceDone,
    QueuedInstance, WorkerStatusBackend, WorkerStatusUpdate,
};

/// Persist runner state and action results in Postgres.
#[derive(Clone)]
pub struct PostgresBackend {
    pool: PgPool,
    query_counts: Arc<Mutex<HashMap<String, usize>>>,
    batch_size_counts: Arc<Mutex<HashMap<String, HashMap<usize, usize>>>>,
}

impl PostgresBackend {
    pub fn new(pool: PgPool) -> Self {
        Self {
            pool,
            query_counts: Arc::new(Mutex::new(HashMap::new())),
            batch_size_counts: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    #[obs]
    pub async fn connect(dsn: &str) -> BackendResult<Self> {
        let pool = PgPool::connect(dsn).await?;
        db::run_migrations(&pool).await?;
        Ok(Self::new(pool))
    }

    pub fn pool(&self) -> &PgPool {
        &self.pool
    }

    /// Insert queued instances for run-loop consumption.
    #[obs]
    pub async fn queue_instances(&self, instances: &[QueuedInstance]) -> BackendResult<()> {
        if instances.is_empty() {
            return Ok(());
        }
        let mut queued_payloads = Vec::new();
        let mut runner_payloads = Vec::new();
        for instance in instances {
            queued_payloads.push((instance.instance_id, Self::serialize(instance)?));
            let graph = build_graph_update(instance, instance.instance_id);
            runner_payloads.push((
                instance.instance_id,
                instance.entry_node,
                Self::serialize(&graph)?,
            ));
        }

        let mut queued_builder: QueryBuilder<Postgres> =
            QueryBuilder::new("INSERT INTO queued_instances (instance_id, payload) ");
        queued_builder.push_values(queued_payloads.iter(), |mut builder, (id, payload)| {
            builder.push_bind(*id).push_bind(payload.as_slice());
        });

        let mut runner_builder: QueryBuilder<Postgres> =
            QueryBuilder::new("INSERT INTO runner_instances (instance_id, entry_node, state) ");
        runner_builder.push_values(
            runner_payloads.iter(),
            |mut builder, (id, entry, payload)| {
                builder
                    .push_bind(*id)
                    .push_bind(*entry)
                    .push_bind(payload.as_slice());
            },
        );

        let mut tx = self.pool.begin().await?;
        Self::count_query(&self.query_counts, "insert:queued_instances");
        Self::count_batch_size(
            &self.batch_size_counts,
            "insert:queued_instances",
            instances.len(),
        );
        queued_builder.build().execute(&mut *tx).await?;
        Self::count_query(&self.query_counts, "insert:runner_instances");
        Self::count_batch_size(
            &self.batch_size_counts,
            "insert:runner_instances",
            instances.len(),
        );
        runner_builder.build().execute(&mut *tx).await?;
        tx.commit().await?;
        Ok(())
    }

    /// Delete all queued instances from the backing table.
    #[obs]
    pub async fn clear_queue(&self) -> BackendResult<()> {
        Self::count_query(&self.query_counts, "delete:queued_instances_all");
        sqlx::query("DELETE FROM queued_instances")
            .execute(&self.pool)
            .await?;
        Ok(())
    }

    /// Delete all persisted runner data for a clean benchmark run.
    #[obs]
    pub async fn clear_all(&self) -> BackendResult<()> {
        Self::count_query(&self.query_counts, "truncate:runner_tables");
        sqlx::query(
            r#"
            TRUNCATE runner_graph_updates,
                     runner_actions_done,
                     runner_instances,
                     runner_instances_done,
                     queued_instances
            RESTART IDENTITY
            "#,
        )
        .execute(&self.pool)
        .await?;
        Ok(())
    }

    pub fn query_counts(&self) -> HashMap<String, usize> {
        self.query_counts
            .lock()
            .expect("query counts poisoned")
            .clone()
    }

    pub fn batch_size_counts(&self) -> HashMap<String, HashMap<usize, usize>> {
        self.batch_size_counts
            .lock()
            .expect("batch size counts poisoned")
            .clone()
    }

    fn count_query(counts: &Arc<Mutex<HashMap<String, usize>>>, label: &str) {
        let mut guard = counts.lock().expect("query counts poisoned");
        *guard.entry(label.to_string()).or_insert(0) += 1;
    }

    fn count_batch_size(
        counts: &Arc<Mutex<HashMap<String, HashMap<usize, usize>>>>,
        label: &str,
        size: usize,
    ) {
        if size == 0 {
            return;
        }
        let mut guard = counts.lock().expect("batch size counts poisoned");
        let entry = guard.entry(label.to_string()).or_default();
        *entry.entry(size).or_insert(0) += 1;
    }

    fn serialize<T: serde::Serialize>(value: &T) -> Result<Vec<u8>, BackendError> {
        rmp_serde::to_vec_named(value).map_err(|e| BackendError::Message(e.to_string()))
    }

    fn deserialize<T: serde::de::DeserializeOwned>(payload: &[u8]) -> Result<T, BackendError> {
        rmp_serde::from_slice(payload).map_err(|e| BackendError::Message(e.to_string()))
    }

    /// Upsert worker status for monitoring and activity graphs.
    #[obs]
    pub async fn upsert_worker_status(&self, status: &WorkerStatusUpdate) -> BackendResult<()> {
        Self::count_query(&self.query_counts, "upsert:worker_status");
        sqlx::query(
            r#"
            INSERT INTO worker_status (
                pool_id,
                worker_id,
                throughput_per_min,
                total_completed,
                last_action_at,
                updated_at,
                median_dequeue_ms,
                median_handling_ms,
                dispatch_queue_size,
                total_in_flight,
                active_workers,
                actions_per_sec,
                median_instance_duration_secs,
                active_instance_count,
                total_instances_completed,
                instances_per_sec,
                instances_per_min,
                time_series
            )
            VALUES ($1, 0, $2, $3, $4, NOW(), $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)
            ON CONFLICT (pool_id, worker_id)
            DO UPDATE SET
                throughput_per_min = EXCLUDED.throughput_per_min,
                total_completed = EXCLUDED.total_completed,
                last_action_at = EXCLUDED.last_action_at,
                updated_at = EXCLUDED.updated_at,
                median_dequeue_ms = EXCLUDED.median_dequeue_ms,
                median_handling_ms = EXCLUDED.median_handling_ms,
                dispatch_queue_size = EXCLUDED.dispatch_queue_size,
                total_in_flight = EXCLUDED.total_in_flight,
                active_workers = EXCLUDED.active_workers,
                actions_per_sec = EXCLUDED.actions_per_sec,
                median_instance_duration_secs = EXCLUDED.median_instance_duration_secs,
                active_instance_count = EXCLUDED.active_instance_count,
                total_instances_completed = EXCLUDED.total_instances_completed,
                instances_per_sec = EXCLUDED.instances_per_sec,
                instances_per_min = EXCLUDED.instances_per_min,
                time_series = EXCLUDED.time_series
            "#,
        )
        .bind(status.pool_id)
        .bind(status.throughput_per_min)
        .bind(status.total_completed)
        .bind(status.last_action_at)
        .bind(status.median_dequeue_ms)
        .bind(status.median_handling_ms)
        .bind(status.dispatch_queue_size)
        .bind(status.total_in_flight)
        .bind(status.active_workers)
        .bind(status.actions_per_sec)
        .bind(status.median_instance_duration_secs)
        .bind(status.active_instance_count)
        .bind(status.total_instances_completed)
        .bind(status.instances_per_sec)
        .bind(status.instances_per_min)
        .bind(&status.time_series)
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    #[obs]
    async fn save_graphs_impl(&self, graphs: &[GraphUpdate]) -> BackendResult<()> {
        if graphs.is_empty() {
            return Ok(());
        }
        Self::count_query(&self.query_counts, "update:runner_instances_state");
        Self::count_batch_size(
            &self.batch_size_counts,
            "update:runner_instances_state",
            graphs.len(),
        );
        let mut payloads = Vec::with_capacity(graphs.len());
        for graph in graphs {
            payloads.push((graph.instance_id, Self::serialize(graph)?));
        }
        let mut builder: QueryBuilder<Postgres> =
            QueryBuilder::new("UPDATE runner_instances AS ri SET state = v.state FROM (");
        builder.push_values(payloads.iter(), |mut b, (instance_id, payload)| {
            b.push_bind(*instance_id).push_bind(payload.as_slice());
        });
        builder.push(") AS v(instance_id, state) WHERE ri.instance_id = v.instance_id");
        builder.build().execute(&self.pool).await?;
        Ok(())
    }

    #[obs]
    async fn save_actions_done_impl(&self, actions: &[ActionDone]) -> BackendResult<()> {
        if actions.is_empty() {
            return Ok(());
        }
        Self::count_query(&self.query_counts, "insert:runner_actions_done");
        Self::count_batch_size(
            &self.batch_size_counts,
            "insert:runner_actions_done",
            actions.len(),
        );
        let mut payloads = Vec::new();
        for action in actions {
            payloads.push((
                action.node_id,
                action.action_name.clone(),
                action.attempt,
                Self::serialize(&action.result)?,
            ));
        }
        let mut builder: QueryBuilder<Postgres> = QueryBuilder::new(
            "INSERT INTO runner_actions_done (node_id, action_name, attempt, result) ",
        );
        builder.push_values(
            payloads.iter(),
            |mut b, (node_id, name, attempt, payload)| {
                b.push_bind(*node_id)
                    .push_bind(name)
                    .push_bind(*attempt)
                    .push_bind(payload.as_slice());
            },
        );
        builder.build().execute(&self.pool).await?;
        Ok(())
    }

    #[obs]
    async fn get_queued_instances_impl(&self, size: usize) -> BackendResult<Vec<QueuedInstance>> {
        if size == 0 {
            return Ok(Vec::new());
        }
        let mut tx = self.pool.begin().await?;
        Self::count_query(&self.query_counts, "select:queued_instances");
        let rows = sqlx::query(
            "SELECT instance_id, payload FROM queued_instances ORDER BY created_at LIMIT $1 FOR UPDATE SKIP LOCKED",
        )
        .bind(size as i64)
        .fetch_all(&mut *tx)
        .await?;
        if rows.is_empty() {
            tx.commit().await?;
            return Ok(Vec::new());
        }
        Self::count_batch_size(
            &self.batch_size_counts,
            "select:queued_instances",
            rows.len(),
        );
        let ids: Vec<Uuid> = rows.iter().map(|row| row.get::<Uuid, _>(0)).collect();
        Self::count_query(&self.query_counts, "delete:queued_instances_by_id");
        sqlx::query("DELETE FROM queued_instances WHERE instance_id = ANY($1)")
            .bind(&ids)
            .execute(&mut *tx)
            .await?;
        tx.commit().await?;

        let mut instances = Vec::new();
        for row in rows {
            let instance_id: Uuid = row.get(0);
            let payload: Vec<u8> = row.get(1);
            let mut instance: QueuedInstance = Self::deserialize(&payload)?;
            instance.instance_id = instance_id;
            instances.push(instance);
        }
        Ok(instances)
    }

    #[obs]
    async fn save_instances_done_impl(&self, instances: &[InstanceDone]) -> BackendResult<()> {
        if instances.is_empty() {
            return Ok(());
        }
        Self::count_query(&self.query_counts, "update:runner_instances_result");
        Self::count_batch_size(
            &self.batch_size_counts,
            "update:runner_instances_result",
            instances.len(),
        );
        let mut payloads = Vec::with_capacity(instances.len());
        for instance in instances {
            let result = match &instance.result {
                Some(value) => Some(Self::serialize(value)?),
                None => None,
            };
            let error = match &instance.error {
                Some(value) => Some(Self::serialize(value)?),
                None => None,
            };
            payloads.push((instance.executor_id, result, error));
        }
        let mut builder: QueryBuilder<Postgres> = QueryBuilder::new(
            "UPDATE runner_instances AS ri SET result = v.result, error = v.error FROM (",
        );
        builder.push_values(payloads.iter(), |mut b, (instance_id, result, error)| {
            b.push_bind(*instance_id)
                .push_bind(result.as_deref())
                .push_bind(error.as_deref());
        });
        builder.push(") AS v(instance_id, result, error) WHERE ri.instance_id = v.instance_id");
        builder.build().execute(&self.pool).await?;
        Ok(())
    }
}

impl BaseBackend for PostgresBackend {
    fn clone_box(&self) -> Box<dyn BaseBackend> {
        Box::new(self.clone())
    }

    fn save_graphs<'a>(&'a self, graphs: &'a [GraphUpdate]) -> BoxFuture<'a, BackendResult<()>> {
        Box::pin(self.save_graphs_impl(graphs))
    }

    fn save_actions_done<'a>(
        &'a self,
        actions: &'a [ActionDone],
    ) -> BoxFuture<'a, BackendResult<()>> {
        Box::pin(self.save_actions_done_impl(actions))
    }

    fn get_queued_instances<'a>(
        &'a self,
        size: usize,
    ) -> BoxFuture<'a, BackendResult<Vec<QueuedInstance>>> {
        Box::pin(self.get_queued_instances_impl(size))
    }

    fn save_instances_done<'a>(
        &'a self,
        instances: &'a [InstanceDone],
    ) -> BoxFuture<'a, BackendResult<()>> {
        Box::pin(self.save_instances_done_impl(instances))
    }
}

impl WorkerStatusBackend for PostgresBackend {
    fn upsert_worker_status<'a>(
        &'a self,
        status: &'a WorkerStatusUpdate,
    ) -> BoxFuture<'a, BackendResult<()>> {
        Box::pin(async move { PostgresBackend::upsert_worker_status(self, status).await })
    }
}

fn build_graph_update(instance: &QueuedInstance, instance_id: Uuid) -> GraphUpdate {
    if let Some(state) = &instance.state {
        return GraphUpdate {
            instance_id,
            nodes: state.nodes.clone(),
            edges: state.edges.clone(),
        };
    }
    let nodes = instance
        .nodes
        .clone()
        .expect("queued instance missing nodes");
    let edges = instance
        .edges
        .clone()
        .expect("queued instance missing edges");
    GraphUpdate {
        instance_id,
        nodes,
        edges,
    }
}
